### PR TITLE
CI: Overhaul static checks to use `pre-commit`

### PR DESCRIPTION
**Pull request by Repiteo**
24-05-05 13-12-32 chrome[^1]

Consolidates (almost) every single `static_checks` function/command into `pre-commit` directly. This has the benefit of adding these static checks as proper pre-commit hooks to the repo itself, so more issues have the potential to be caught well ahead of time. If possible, existing `misc` scripts were either replaced with a more parsable python equivalent, or removed entirely if a hook repo exists to achieve the same goal.

Only three exceptions remain that are still checked via `static_checks`: `gitignore_check.sh`, `pytest`, and `xmllint`. `gitignore_check.sh` is pretty specific to the workflow of that file specifically, so it's most sensible to keep as-is. `xmllint` technically *has* a hook available, but it requires docker & that's not **nearly** as likely for a user to have compared to npm/dotnet. `pytest` actually would've worked just fine, but based off a conversation[^2] when I attempted to add it to pre-commit, it isn't a good candidate for a pre-commit so I'll leave it as-is.

[^1]: https://github.com/godotengine/godot/pull/91417#pullrequestreview-2034011566
[^2]: https://github.com/pre-commit/pre-commit.com/pull/960

### CODE REVIEW

**Comment by akien-mga** [comment]
Thanks for working on this! That's amazing.

I have yet to review the code properly, but some minor issues found when testing on Linux with `pre-commit run --all`:

- `file_format` seems to find a bunch of stuff to fix that we missed in the `master` branch. I made https://github.com/godotengine/godot/pull/91696 with those changes.

- The `eslint-html` check errors out:
```
eslint-html..............................................................Failed
- hook id: eslint
- exit code: 2

Oops! Something went wrong! :(

ESLint: 8.46.0

ESLint couldn't find the config "airbnb-base" to extend from. Please check that the name of the config is correct.

The config "airbnb-base" was referenced from the config file in "/home/akien/Godot/godot/platform/web/.eslintrc.js".

If you still have problems, please stop by https://eslint.org/chat/help to chat with the team.
```

File: .github/workflows/static_checks.yml [hunk 1]
```diff
@@ -49,51 +34,12 @@ jobs:
...
         run: |
           xmllint --noout --schema doc/class.xsd doc/classes/*.xml modules/*/doc_classes/*.xml platform/*/doc_classes/*.xml
-
-      - name: Documentation checks
-        run: |
-          doc/tools/doc_status.py doc/classes modules/*/doc_classes platform/*/doc_classes
-
-      - name: Spell checks via codespell
-        if: github.event_name == 'pull_request' && env.CHANGED_FILES != ''
-        uses: codespell-project/actions-codespell@v2
```
Comment (akien-mga) [inline .github/workflows/static_checks.yml:95]: Just noting that we'll miss one feature here, which is that this action provides a problem matcher in the GitHub diff view.

But in my experience most contributors don't seem to notice it (it's only visible in diff view, not on the PR itself), and the action was lacking in other configuration aspects so I think the tradeoffs are fine.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -19,18 +23,119 @@ repos:
...
+    rev: v8.46.0
+    hooks:
+    - id: eslint
+      name: eslint-engine
+      files: ^(platform/web/js/engine/|js/jsdoc2rst/).*\.js$
+      args: [--fix, --no-eslintrc, --config, platform/web/.eslintrc.engine.js]
+      additional_dependencies:
+      - eslint@8.46.0
+      - eslint-config-airbnb-base@15.0.0
+      - eslint-plugin-import@2.28.0
```
Comment (akien-mga) [inline .pre-commit-config.yaml:90]: Shouldn't these be indented? Same below.
```suggestion
      additional_dependencies:
        - eslint@8.46.0
        - eslint-config-airbnb-base@15.0.0
        - eslint-plugin-import@2.28.0
```

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -19,18 +23,119 @@ repos:
...
+  #     - id: xmllint
+  #       language: docker
+  #       types_or: [text]
+  #       files: ^(doc/classes|.*/doc_classes)/.*\.xml$
+  #       args: [--schema, doc/class.xsd]
+
+  - repo: https://github.com/pre-commit/mirrors-eslint
+    rev: v8.46.0
+    hooks:
+    - id: eslint
```
Comment (akien-mga) [inline .pre-commit-config.yaml:83]: All these hooks have the same `id`, is that intended?

**Review by akien-mga — APPROVED** [review]
Reviewed the code, looks really good.

**Comment by akien-mga** [comment]
BTW, if you want to have a look at how we could solve the problem those two PRs aimed to address (users not having `python3` in PATH): 
- #90634
- #90662

File: .github/workflows/static_checks.yml (hunk 1 above)
Comment (Repiteo) [reply .github/workflows/static_checks.yml:95]: pre-commit should be set to output a git-diff log if any checks fail or changes are made, so it won't be that much of a loss.

**Comment by Repiteo** [comment]
I'll leave the `python3` as-is for now then, and do some digging on how it's implemented in `pre-commit`. I *think* it can be setup such that it won't rely on shbang whatsoever, but I'm fine tackling that in a followup & letting this merge as-is.

**Comment by akien-mga** [comment]
Beautiful:
```
clang-format.............................................................Passed
black....................................................................Passed
mypy.....................................................................Passed
codespell................................................................Passed
eslint-engine............................................................Passed
eslint-libs..............................................................Passed
eslint-sw................................................................Passed
eslint-html..............................................................Passed
make-rst.................................................................Passed
doc-status...............................................................Passed
jsdoc....................................................................Passed
copyright-headers........................................................Passed
header-guards............................................................Passed
file-format..............................................................Passed
dotnet-format............................................................Passed
```

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! Great refactor :tada:

**Comment by AThousandShips** [comment]
It looks like this broke CI and no style checks are run at all, see:
https://github.com/godotengine/godot/actions/runs/9043846245/job/24851885660?pr=91829#step:6:82

Which has changed files but skips them

**Comment by Repiteo** [comment]
Then it probably needs that git diff logic passed to it after all, or something similar. Should be an easy fix either way, I'll get something up later today

**Comment by patwork** [comment]
@Repiteo Hi, would you be so kind as to take another look at `.pre-commit-config.yaml`? I've been working on upgrading eslint to version 9 and I see now that the automatic checks are returning errors:

https://github.com/godotengine/godot/pull/91863/checks

- eslint-* because pre-commit is using eslint 8
- jsdoc propably because some arguments are missing?

Is it possible for pre-commit to run `npm run docs` and `npm run lint` directly from `/platform/web`? Especially for eslint this is important, due to the fact that for version 9 you have to change the working directory first (the npm script does this).

**Pull request merged**